src/rows.ts

```diff
@@ -39,62 +38,62 @@ export class Rows {
   }
 
   /**
-   * Returns a readable stream of {@link Row}s for responses
+   * Returns an async iterator of {@link Row}s for responses
    * in {@link StreamableDataFormat} format.
    *
-   * The method will throw if called on a response in non-streamable format,
-   * and if the underlying stream was already consumed
-   * by calling the other methods
+   * If selected format is non-streamable,
+   * or the underlying stream was already consumed,
+   * it will throw when accessing the next element
    */
-  stream(): Stream.Readable {
-    // If the underlying stream has already ended by calling `text` or `json`,
-    // Stream.pipeline will create a new empty stream
-    // but without "readableEnded" flag set to true
+  async *stream(): AsyncGenerator<Row, void> {
     if (this._stream.readableEnded) {
       throw Error(streamAlreadyConsumedMessage)
     }
-
     validateStreamFormat(this.format)
+    const textDecoder = new TextDecoder()
+    let decodedChunk = ''
+    for await (const chunk of this._stream) {
+      decodedChunk += textDecoder.decode(chunk, { stream: true })
+      let idx = 0
+      while (true) {
+        idx = decodedChunk.indexOf('\n')
+        if (idx !== -1) {
+          const line = decodedChunk.slice(0, idx)
+          decodedChunk = decodedChunk.slice(idx + 1)
+          yield {
+            /**
+             * Returns a string representation of a row.
+             */
+            text(): string {
+              return line
+            },
 
-    return Stream.pipeline(
-      this._stream,
-      // only JSON-based format are supported at the moment
-      split((row: string) => new Row(row, 'JSON')),
-      function pipelineCb(err) {
-        if (err) {
-          console.error(err)
+            /**
+             * Returns a JSON representation of a row.
+             * The method will throw if called on a response in JSON incompatible format.
+             *
+             * It is safe to call this method multiple times.
+             */
+            json<T>(): T {
+              return decode(line, 'JSON')
+            },
+          }
+        } else {
+          break
         }
       }
-    )
+    }
+    textDecoder.decode() // flush
```

Comment:
Shouldn't it be consumed? The method typings shows it as
```js
    /**
     * Returns the result of running encoding's decoder. The method can be invoked zero or more times with options's stream set to true, and then once without options's stream (or set to false), to process a fragmented input. If the invocation without options's stream (or set to false) has no input, it's clearest to omit both arguments.
     *
     * ```
     * var string = "", decoder = new TextDecoder(encoding), buffer;
     * while(buffer = next_chunk()) {
     *   string += decoder.decode(buffer, {stream:true});
     * }
     * string += decoder.decode(); // end-of-queue
     * ```
     *
     * If the error mode is "fatal" and encoding's decoder returns error, throws a TypeError.
     */
    decode(input?: BufferSource, options?: TextDecodeOptions): string;
```